### ISSUE TITLE
does not play audio but logs "Track is not playable"

### ISSUE THREAD

**Issue by mgoltzsche** [natumbri/mopidy-youtube#241]
@natumbri First of all thanks for maintaining this great plugin!

I am running mopidy-youtube (latest installed via pip) within an alpine:3.17-based [docker container](https://github.com/mgoltzsche/mopidy-container/tree/v0.5.0) with yt-dlp (2023.03.04) and, after specifying the `youtube_api_key`, I can search Youtube but, when clicking a track, it does not play audio but logs:
```
[youtube] Extracting URL: https://www.youtube.com/watch?v=95G-PW0YdaI
[youtube] 95G-PW0YdaI: Downloading webpage
[youtube] 95G-PW0YdaI: Downloading android player API JSON
[info] 95G-PW0YdaI: Downloading 1 format(s): 251
[dashsegments] Total fragments: 1
[download] Destination: /var/lib/mopidy/cache/youtube/95G-PW0YdaI.webm
ERROR    2023-05-06 17:05:11,314 [30:YouTubeCoreListener-17 (_actor_loop)] mopidy_youtube
  audio_url error 'total_bytes' (videoId: 95G-PW0YdaI)
WARNING  2023-05-06 17:05:11,314 [30:Core-14 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: youtube:video:95G-PW0YdaI
```
Looking into the file system, I can see apparently the first fragment of the track has been downloaded but the final audio file is empty:
```
$ sudo ls -lah /var/lib/kubemate/k3s/storage/pvc-82cbed86-b86c-4684-9eb3-2ec6a3d10692_default_mopidy-data/cache/youtube/
insgesamt 248K
drwxr-xr-x 2 systemd-network 18 4,0K Mai  6 19:05 .
drwxr-xr-x 3 systemd-network 18 4,0K Mai  6 19:04 ..
-rw-r--r-- 1 systemd-network 18 107K Mai  6 19:05 95G-PW0YdaI.jpg
-rw-r--r-- 1 systemd-network 18    0 Mai  6 19:05 95G-PW0YdaI.webm
-rw-r--r-- 1 systemd-network 18 127K Mai  6 19:05 95G-PW0YdaI.webm-Frag1
-rw-r--r-- 1 systemd-network 18   50 Mai  6 19:05 95G-PW0YdaI.webm.ytdl
```

The fact that the final audio file is empty seems to be causing the mopidy-youtube plugin to fail with `audio_url error 'total_bytes'`.
I verified that the downloaded audio fragment (*.webm-Frag1) is playable using another player.
This is not specific to a particular track but seems to happen with any track.

I suspect an incompatibility between youtube-music and yt-dlp or rather some version misalignment.

Do you know what could be causing this and how to fix it by any chance?

**Comment by natumbri** [comment]
Thanks for the feedback.  I think there is something broken in the code that starts playback before a file is completely downloaded.  I haven't fixed it, but if you comment it out, it should work (albeit with a delay before starting playback).

This is the relevant part of the code: https://github.com/natumbri/mopidy-youtube/blob/0a069b75716bddfc849492c3df9f450fa4563c61/mopidy_youtube/youtube.py#L413-L439

Or, that may not be the problem, at all!  In which case, please do let me know.

Cheers,
Nik

**Comment by mgoltzsche** [comment]
[Commenting that code block out](https://github.com/mgoltzsche/mopidy-youtube/commit/1091d3b7de79f3c9b1c8ff8e5f670f91896f0fa2) makes it play audio, indeed.
Thanks @natumbri!

**Comment by natumbri** [comment]
OK, I think I've fixed this properly.  New release pushed.  Reopen if the problem comes back.  Thanks!

**Comment by mgoltzsche** [comment]
I can confirm it works now. Thanks Nik!

**Comment by mgoltzsche** [comment]
Unfortunately, your last release didn't fix the issue or it caused a new one. I must have tested that wrongly previously, without clearing the cache beforehand. [With your new release](https://github.com/mgoltzsche/mopidy-container/releases/tag/v0.6.1) my mopidy container dies right after downloading a file but without logging any error (that may be because a child process would have logged the error later but when the parent dies the container terminates). For now the fix for me was to [downgrade to my previous mopidy container](https://github.com/mgoltzsche/mopidy-container/releases/tag/v0.6.0) version that was not using your last release but my fork. I'll investigate some other day why the mopidy container crashes with your latest version. (I remember it downloaded at least more than 2% before it died if that helps).

**Comment by natumbri** [comment]
Sorry to hear this! It is working fine for me, so it is a bit hard to work out what is going on. It is a shame I haven't made a proper set of automated tests to test it all more rigorously.

Could it be a permissions problem?

**Comment by mgoltzsche** [comment]
Turns out the problem also appeared with the previous version and is not related to this issue.

**TL;DR:**
Apparently the reason is that Kubernetes restarts the mopidy container since its readiness/liveness probe (health check) times out waiting for a response when mopidy is downloading a track from youtube when the track is very long/large. Looks like the download blocks the entire mopidy server process. I also noticed that behaviour within the GUI where the volume cannot be changed after mopidy-youtube initiated a download and before the download completed.
The solution for me was to [increase the readiness/liveness probe timeout/threshold](https://github.com/mgoltzsche/mopidy-container/releases/tag/v0.6.2) for the mopidy container. (Also, I had to increase the memory and cpu limits because it seems that at some point a whole track is loaded into memory and sometimes there was a bit of stuttering initially.)

**Comment by natumbri** [comment]
Thanks for persisting, and glad to hear you've sorted it out. That mopidy-youtube behaviour doesn't sound ideal (or even necessary) - I wonder if I can do better... Something to think about.

Cheers
Nik
